x-pack/plugins/profiling/server/routes/functions.ts

```diff
@@ -21,59 +23,65 @@ import {
-    eventsIndex,
-    filter
-  );
+async function queryTopNFunctions({
+  logger,
+  client,
+  index,
+  filter,
+  startIndex,
+  endIndex,
+  sampleSize,
+}: {
+  logger: Logger;
+  client: ProfilingESClient;
+  index: string;
+  filter: ProjectTimeQuery;
+  startIndex: number;
+  endIndex: number;
+  sampleSize: number;
+}): Promise<any> {
+  return withProfilingSpan('query_topn_functions', async () => {
+    const eventsIndex = await findDownsampledIndex({ logger, client, index, filter, sampleSize });
 
-  // Manual downsampling if totalCount exceeds sampleSize by 10%.
-  if (totalCount > sampleSize * 1.1) {
-    let downsampledTotalCount = totalCount;
-    const p = sampleSize / totalCount;
-    logger.info('downsampling events with p=' + p);
-    await logExecutionLatency(logger, 'downsampling events', async () => {
-      downsampledTotalCount = downsampleEventsRandomly(stackTraceEvents, p, filter.toString());
+    const { totalCount, stackTraceEvents } = await searchEventsGroupByStackTrace({
+      logger,
+      client,
+      index: eventsIndex,
+      filter,
     });
-    logger.info('downsampled total count: ' + downsampledTotalCount);
-    logger.info('unique downsampled stacktraces: ' + stackTraceEvents.size);
-  }
 
-  const { stackTraces, stackFrameDocIDs, executableDocIDs } = await mgetStackTraces(
-    logger,
-    client,
-    stackTraceEvents
-  );
+    // Manual downsampling if totalCount exceeds sampleSize by 10%.
+    if (totalCount > sampleSize * 1.1) {
+      let downsampledTotalCount = totalCount;
+      const p = sampleSize / totalCount;
+      logger.info('downsampling events with p=' + p);
+      await logExecutionLatency(logger, 'downsampling events', async () => {
+        downsampledTotalCount = downsampleEventsRandomly(stackTraceEvents, p, filter.toString());
+      });
+      logger.info('downsampled total count: ' + downsampledTotalCount);
+      logger.info('unique downsampled stacktraces: ' + stackTraceEvents.size);
+    }
 
-  return Promise.all([
-    mgetStackFrames(logger, client, stackFrameDocIDs),
-    mgetExecutables(logger, client, executableDocIDs),
-  ]).then(([stackFrames, executables]) => {
-    return createTopNFunctions(
-      stackTraceEvents,
-      stackTraces,
-      stackFrames,
-      executables,
-      startIndex,
-      endIndex
+    const { stackTraces, stackFrameDocIDs, executableDocIDs } = await mgetStackTraces({
+      logger,
+      client,
+      events: stackTraceEvents,
+    });
+
+    return withProfilingSpan('mget_stack_frames_and_executables', () =>
+      Promise.all([
+        mgetStackFrames({ logger, client, stackFrameIDs: stackFrameDocIDs }),
+        mgetExecutables({ logger, client, executableIDs: executableDocIDs }),
+      ]).then(([stackFrames, executables]) => {
+        return createTopNFunctions(
+          stackTraceEvents,
+          stackTraces,
+          stackFrames,
+          executables,
+          startIndex,
+          endIndex
+        );
+      })
```

Comment:
I see this code several times. Maybe it's a good opportunity to de-duplicate it when touching it anyways !?

Reply:
Aye, I will look into possibly deduplicating this.